### PR TITLE
do a cyclic scan for orphan objects in zstd memory pool

### PR DESCRIPTION
**Pull request by BrainSlayer**
in non regular usecases allocated memory might stay persistent
in memory pool. this small patch checks every minutes if there are
old objects which can be released from memory pool.

right now with regular use, the pool is checked for old objects on each allocation attempt from this pool. so basicly polling by its use. now consider what happens if someone writes alot of files and stops use of the volume or even unmounts it. so the code will no longer check if objects can be released from the pool. so what has been allocated still stays in pool cache. this is no big issue for common use. but someone discovered this issue while doing tests. personally i know this behaviour and i'm aware of it. its no big issue. just a enhancement. see also #10938

covers https://github.com/openzfs/zfs/issues/10938

Signed-off-by: Sebastian Gottschall <s.gottschall@dd-wrt.com>

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [x] I have run the ZFS Test Suite with this change applied.
- [ ] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

**Comment by mschilli87** [comment]
@BrainSlayer:
> in non regular usecases allocated memory might stay persistent in memory pool.

Why is that? Do we know? Don't we care?


> this small patch checks every minutes if there are old objects which can be released from memory pool.

This seems a bit hacky IMHO. Wouldn't the proper fix be to investigate why the memory ends up orphaned in the first place and ensure it doesn't? 
If you want this as a fallback safeguard for practical reasons, I think it should at least issue a warning when it has to release memory 'forgotten' previously. Otherwise this could end up masking more severe future issues with 'proper' memory release.

**Comment by BrainSlayer** [comment]
@mschilli87 let me explain. right now with regular use, the pool is checked for old objects on each allocation attempt from this pool. so basicly polling by its use. now consider what happens if someone writes alot of files and stops use of the volume or even unmounts it. so the code will no longer check if objects can be released from the pool. so what has been allocated still stays in pool cache. this is no big issue for common use. but someone discovered this issue while doing tests. personally i know this behaviour and i'm aware of it. its no big issue. just a enhancement. see also https://github.com/openzfs/zfs/issues/10938

**Comment by mschilli87** [comment]
@BrainSlayer: Thank you for the explanation. My two cents:

1. This explanation should be in the commit message to document the need for that change.
2. I'd still prefer a fix more closer to the root cause. Why not trigger this check when umounting a volume instead of polling regularly in case it got unmounted? Do we keep pooling unused volumes even after freeing the leftover memory? Or could we stop considering this volume after dealing with it once (within a minute after stopping to use it) and add it back to our list once it get's used again (e.g. in the polling by usage code you mentioned)?

**Comment by BrainSlayer** [comment]
copied the my explaination to the title. will add it later to the commit message after testrun is done

regarding the check procedure. zstd is just a algorithm. it doesnt know of any higher level operation and does not know if a volume is in use, mounted or unmounted. the pool is also not used per volume, but used for the algorithm which can be used by multiple volumes of course. and yes the polling remains also if all is freed. but it only happens once in a minute and does not cause any measureable performance impact since the code which is executed is very small for checking. forget about thinking in volumes. just think about a compression algorithm which is available globally for everything you do. the case i mentioned was just a example. if a any volume is in use and you just dont write or isnt in use, the preallocated ram stays allocated without that patch. so its basicly just a optimization to reduce memory usage if everything related to zstd is in idle state

**Comment by mschilli87** [comment]
That's a good enough explanation for me, as long as it makes it into the commit message and/or a comment in the source code itself. Thanks for taking the time to walk me through.

File: module/zstd/zfs_zstd.c [hunk 1]
```diff
@@ -704,12 +719,18 @@ zstd_init(void)
 		kstat_install(zstd_ksp);
 	}
 
+	/* check if objects in memory pool can be released */
```
Comment (Ornias1993) [inline module/zstd/zfs_zstd.c:722]:
```suggestion
	/* check every 60 seconds if objects in memory pool can be released */
```

**Review by Ornias1993 — APPROVED** [review]
Looks good with a slight nitpick on the note not refering to the 60 seconds.

Ping for @c0d3z3r0 and @allanjude to check this out if they want.

File: module/zstd/zfs_zstd.c [hunk 2]
```diff
@@ -688,6 +690,19 @@ zstd_mempool_deinit(void)
...
+zstd_memory_liberator_cb(void *arg, zthr_t *zthr)
+{
+	/*
+	 * calling alloc with zero size seeks
+	 * and releases orphan objects
+	 */
+	zstd_mempool_alloc(zstd_mempool_cctx, 0);
+	zstd_mempool_alloc(zstd_mempool_dctx, 0);
+	return (B_FALSE);
+}
```
Comment (behlendorf) [inline module/zstd/zfs_zstd.c:703]: Instead of moving the zthr code and creating another thread how about leveraging the existing `arc_reap` thread which already performs this same function.  This thread calls the `arc_kmem_reap_soon()` function which reaps unused objects from a variety of ZFS related caches.  You'd just need to add a `zstd_cache_reap_now()` function and call it.

**Comment by richardelling** [comment]
ISTM that this is the wrong approach. We know when an objectset is removed from the system (umount, destroy, close) and we already have async clean to clear its data from ARC. Why not just extend it to also clear out the zstd memory pool?

**Comment by allanjude** [comment]
> 
> 
> ISTM that this is the wrong approach. We know when an objectset is removed from the system (umount, destroy, close) and we already have async clean to clear its data from ARC. Why not just extend it to also clear out the zstd memory pool?

This is a pool of memory allocations used for compression. It is not tied to any specific objset. It is just that after a period of high activity, the amount of memory used by this pool may be high, and it doesn't shrink when there is no activity.

**Comment by mschilli87** [comment]
@allanjude: That is what confused me at well until @BrainSlayer explained it.
@richardelling: Good to know I am not alone. :wink:

I think this whole discussion just stresses that the change suggested here should come with good documentation.

File: module/zstd/zfs_zstd.c (hunk 2 above)
Comment (BrainSlayer) [reply module/zstd/zfs_zstd.c:703]: @behlendorf circular dependency. you cannot load zfs without zsstd beeing loaded. this is the same reason why i moved zthr.c. so no call to zzstd can be made from code within zfs

**Comment by BrainSlayer** [comment]
> ISTM that this is the wrong approach. We know when an objectset is removed from the system (umount, destroy, close) and we already have async clean to clear its data from ARC. Why not just extend it to also clear out the zstd memory pool?

circular dependency. no call to zstd code can be made from functions within zfs. zzstd must be loaded before zfs is loaded. such solutions are only possible with installable hooks

File: module/zstd/zfs_zstd.c (hunk 2 above)
Comment (behlendorf) [reply module/zstd/zfs_zstd.c:703]: I'm not sure I follow.   I was suggesting that `arc_kmem_reap_soon()` call down in to the `zzstd` module via a new public zstd function, say `zfs_zstd_cache_reap_now()`.

**Comment by BrainSlayer** [comment]
> @allanjude: That is what confused me at well until @BrainSlayer explained it.
> @richardelling: Good to know I am not alone. 😉
> 
> I think this whole discussion just stresses that the change suggested here should come with good documentation.

according to what he wrote, he did understand very well how it works.

File: module/zstd/zfs_zstd.c (hunk 2 above)
Comment (BrainSlayer) [reply module/zstd/zfs_zstd.c:703]: ahm sorry. yes. now i understand. my fault

**Comment by mschilli87** [comment]
@BrainSlayer 
> according to what he wrote, he did understand very well how it works.

Sorry for the ambiguous tagging: I tagged Allan as I was replying to him talking about Richard's confusion.
It was more of a motivation for Allan help explain this in code comment and/or commit message than an attempt to imply Allan was confused himself. :wink:

File: module/zstd/zfs_zstd.c (hunk 2 above)
Comment (BrainSlayer) [reply module/zstd/zfs_zstd.c:703]: i have one problem with arc_kmem_reap_soon etc. the rule how arc_cb_reap is called is bad, it exclusive checks rules related to the available arc memory which has nothing todo with a cyclic cleanup procedure. it triggers only when memory is out and under pressure. consider that my release function releases objects which are aged already. so a pure cylic call is all i need. but  arc_kmem_reap_soon will never be called if the system still has memory left

**Comment by BrainSlayer** [comment]
@behlendorf please check the new solution for handling the memory cleanup.

File: module/zstd/zfs_zstd.c (hunk 2 above)
Comment (behlendorf) [reply module/zstd/zfs_zstd.c:702]: Right, I agree it's a bit awkward.  However, I think even calling `zfs_zstd_cache_reap_now()` every second would be pretty reasonable, particularly if the size of the pool were known and we could return immediately when it was empty.  Adding this same size information to the zstd_stats_t kstat would also be really helpful to make easy to check how much memory is really being used here.

Comment (BrainSlayer) [reply module/zstd/zfs_zstd.c:702]: @behlendorf consider that i iterate the pool which is small of course, but i have to call a trylock for each object. so latency is everything here and since aged/outdated mempool objects is defined with "unused since more than 2 minutes) checking every second is not required. of course it would react fast, but it doesnt need so. with normal workflow this polling is also not required since checking for orphan objects is done on each allocation attempt. this cleanup procedure is only required if all io access has been stopped or finished related to zstd compression. so is there any hurry? surelly not.

File: scripts/commitcheck.sh [hunk 3]
```diff
@@ -22,7 +22,7 @@ function test_commit_bodylength()
     length="72"
     body=$(git log -n 1 --pretty=%b "$REF" | grep -Ev "http(s)*://" | grep -E -m 1 ".{$((length + 1))}")
     if [ -n "$body" ]; then
-        echo "error: commit message body contains line over ${length} characters"
+        echo "error: commit message body contains line over ${length} characters ${body}"
```
Comment (behlendorf) [inline scripts/commitcheck.sh:25]: This change looks unrelated.

File: module/zstd/zfs_zstd.c [hunk 4]
```diff
@@ -688,6 +688,19 @@ zstd_mempool_deinit(void)
...
 }
 
+/* release unused orphan memory from pool */
+
+void
+zfs_zstd_cache_reap_now(void)
+{
+	/*
+	 * calling alloc with zero size seeks
+	 * and releases orphan objects
```
Comment (behlendorf) [inline module/zstd/zfs_zstd.c:698]: I don't think "orphan objects" is right phrase to use here since it implies their lost.  How about "unused objected" and moving this all to the block comment above the function.

```c
/*
 * Releases unused memory from the pool by calling
 * zstd_mempool_alloc() with a zero size.
 */
void
zfs_zstd_cache_reap_now(void)
```

File: module/zstd/zfs_zstd.c (hunk 2 above)
Comment (behlendorf) [reply module/zstd/zfs_zstd.c:702]: > this cleanup procedure is only required if all io access has been stopped or finished related to zstd compression

It would potentially be desirable to be able to reclaim this space when free memory is low, but changing the polling frequency of course wouldn't help since it won't age out for 2 minutes.  Let's go ahead and put this reclaim mechanism is place so it will get reclaimed and we can always refine it.

File: module/zfs/arc.c [hunk 5]
```diff
@@ -4996,6 +4998,14 @@ arc_reap_cb_check(void *arg, zthr_t *zthr)
 		arc_no_grow = B_FALSE;
 	}
 
+	/*
+	 * this function should be called unconditional
+	 * every 60 seconds, so we need to use this small little
+	 * quirk to avoid a independend thread.
+	 */
```
Comment (behlendorf) [inline module/zfs/arc.c:5005]: How about:

```c
        /*
         * Called unconditionally every 60 seconds to reclaim unused
         * zstd compression and decompression contexts.  This is done
         * here to avoid the need for an independent thread.
         */
```

**Review by behlendorf — CHANGES_REQUESTED** [review]

File: module/zfs/arc.c (hunk 5 above)
Comment (Ornias1993) [reply module/zfs/arc.c:5005]: Yeah, "should be" is not the right wording indeed.

File: scripts/commitcheck.sh (hunk 3 above)
Comment (BrainSlayer) [reply scripts/commitcheck.sh:25]: ahm yes. debugging leftover

File: module/zstd/zfs_zstd.c (hunk 4 above)
Comment (BrainSlayer) [reply module/zstd/zfs_zstd.c]: done

File: module/zfs/arc.c (hunk 5 above)
Comment (BrainSlayer) [reply module/zfs/arc.c:5005]: done

**Comment by BrainSlayer** [comment]
@behlendorf regarding the way when to reclaim space and when not. the pool allocator is self managing. waiting until memory is low, if avoidable is not a good way from my point of view. in our testcase we had about 1.5 gb unused memory left. thats no small piece. releasing it if not in use for a while avoids such situations which will slow down zfs. so if avoidable without performance impact is a good approach

File: module/zstd/zfs_zstd.c (hunk 2 above)
Comment (BrainSlayer) [reply module/zstd/zfs_zstd.c:702]: regarding the way when to reclaim space and when not. the pool allocator is self managing. waiting until memory is low, if avoidable is not a good way from my point of view. in our testcase we had about 1.5 gb unused memory left. thats no small piece. releasing it if not in use for a while avoids such situations which will slow down zfs. so if avoidable without performance impact is a good approach

**Review by behlendorf — APPROVED** [review]
